pkg/operators/env/env.go

```diff
@@ -0,0 +1,147 @@
+// Copyright 2024-2025 The Inspektor Gadget authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package env
+
+import (
+	"fmt"
+	"os"
+	"slices"
+	"strings"
+
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/datasource"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/gadget-service/api"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/operators"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/params"
+)
+
+const (
+	Name     = "env"
+	Priority = 1
+
+	AnnotationPrefix = "env.fields."
+
+	ParamEnvVars = "env-vars"
+)
+
+type Config struct {
+	AllowedVars []string
+}
+
+type envOperator struct {
+	config Config
+}
+
+func (e *envOperator) Name() string {
+	return Name
+}
+
+func (e *envOperator) Init(params *params.Params) error {
+	e.config.AllowedVars = strings.Split(params.Get(ParamEnvVars).AsString(), ",")
+	return nil
+}
+
+func (e *envOperator) GlobalParams() api.Params {
+	return api.Params{
+		{
+			Key:         "env-vars",
+			Description: "Comma-separated list of environment variables to allow to be included",
+			Title:       "Allowed Environment Variables",
+		},
+	}
+}
+
+func (e *envOperator) InstanceParams() api.Params {
+	return nil
+}
+
+func (e *envOperator) InstantiateDataOperator(gadgetCtx operators.GadgetContext, instanceParamValues api.ParamValues) (operators.DataOperatorInstance, error) {
+	instance := &envOperatorInstance{
+		op:     e,
+		fields: make(map[datasource.DataSource][]datasource.DataFunc),
+	}
+	err := instance.init(gadgetCtx)
+	if err != nil {
+		return nil, err
+	}
+	return instance, nil
+}
+
+func (e *envOperator) Priority() int {
+	return Priority
+}
+
+type envOperatorInstance struct {
+	op     *envOperator
+	fields map[datasource.DataSource][]datasource.DataFunc
+}
+
+func (e *envOperatorInstance) Name() string {
+	return Name
+}
+
+func (e *envOperatorInstance) init(gadgetCtx operators.GadgetContext) error {
+	e.fields = make(map[datasource.DataSource][]datasource.DataFunc)
+	for _, ds := range gadgetCtx.GetDataSources() {
```

Comment:
I think this should be done on the `PreStart` to ensure we are registering to all data sources as operators with lower prio might register data sources during instantiate phase.

Reply:
PreStart is too late. Datasourced and fields should be initialized on instantiation - otherwise it won't get picked up by `GetGadgetInfo`, which means it would be missing from column lists, as options for sorting etcpp.

Reply:
That's what I'm saying. Data sources will be registered in instantiation, but this op might lost those data sources registered on operators with lower priority.

Reply:
It still needs to be done on initialization - we could set priority to a higher number if you think it might conflict, but it still should be done on instantiation so the field ends up in the gadget info.

Reply:
I think we should standardize that. All op register datasources on instantiate and all op get data source son PreStart.

Reply:
> PreStart is too late. 
> 
Why?

Reply:
> > PreStart is too late.
> 
> Why?

Frontends (e.g. cobra, now IG Desktop) rely on the info from `GetGadgetInfo` in order to show which fields there are - in order to sort, filter, etcpp. And `GetGadgetInfo` only instantiates and never calls other lifecycle operations. So with only the instantiation, datasources should have been initialized AND enriched (well, fields added etc.)

Reply:
> I think we should standardize that. All op register datasources on instantiate and all op get data source son PreStart.

Still in progress in #3778 😄 

General rule of thumb currently is: instantiate for DS creation / manipulation, PreStart for subscriptions to not miss events and Start for emitting events.